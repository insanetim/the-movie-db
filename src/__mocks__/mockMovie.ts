import type { IBackdrop, ICast, ICrew, IMovie, IMovieDetailExtended } from 'src/interfaces/movie.interface'

export const mockMovie: IMovie = {
  adult: false,
  backdrop_path: '/image',
  genre_ids: [1, 2, 3],
  id: 123,
  original_language: 'en',
  original_title: 'test/title',
  overview: 'test/overview',
  popularity: 123,
  poster_path: '/image',
  release_date: '01-01-1999',
  title: 'test/title',
  video: false,
  vote_average: 123,
  vote_count: 123
}

export const mockMovieDetail: IMovieDetailExtended = {
  accountStates: { favorite: false, id: 123, rated: false, watchlist: false },
  adult: false,
  backdrop_path: '/image',
  belongs_to_collection: null,
  budget: 1000000,
  credits: {
    cast: [
      {
        adult: true,
        cast_id: 123,
        character: 'test/character',
        credit_id: '123',
        gender: null,
        id: 123,
        known_for_department: '',
        name: 'test/name',
        order: 123,
        original_name: 'test/name',
        popularity: 123,
        profile_path: '/image'
      }
    ],
    crew: [
      {
        adult: true,
        credit_id: '123',
        department: 'test/department',
        gender: null,
        id: 123,
        job: 'test/job',
        known_for_department: '',
        name: 'test/name',
        original_name: 'test/name',
        popularity: 123,
        profile_path: '/image'
      }
    ],
    id: 123
  },
  genre_ids: [1, 2, 3],
  genres: [{ id: 1, name: 'test/genre' }],
  homepage: null,
  id: 123,
  images: [
    {
      aspect_ratio: 0.5625,
      file_path: '/image',
      height: 900,
      iso_639_1: null,
      vote_average: 123,
      vote_count: 123,
      width: 1600
    }
  ],
  imdb_id: null,
  original_language: 'en',
  original_title: 'test/title',
  overview: 'test/overview',
  popularity: 123,
  poster_path: '/image',
  production_companies: [{ id: 123, logo_path: null, name: 'test/company', origin_country: 'USA' }],
  production_countries: [{ iso_3166_1: 'en', name: 'USA' }],
  release_date: '01-01-1999',
  revenue: 3000000,
  runtime: 150,
  spoken_languages: [{ iso_639_1: 'en', name: 'EN' }],
  status: 'test/status',
  tagline: null,
  title: 'test/title',
  video: false,
  vote_average: 123,
  vote_count: 123
}

export const mockCast: ICast = {
  adult: true,
  cast_id: 123,
  character: 'test/character',
  credit_id: '123',
  gender: null,
  id: 123,
  known_for_department: '',
  name: 'test/name',
  order: 123,
  original_name: 'test/name',
  popularity: 123,
  profile_path: '/image'
}

export const mockCrew: ICrew = {
  adult: true,
  credit_id: '123',
  department: 'test/department',
  gender: null,
  id: 123,
  job: 'test/job',
  known_for_department: '',
  name: 'test/name',
  original_name: 'test/name',
  popularity: 123,
  profile_path: '/image'
}

export const mockImage: IBackdrop = {
  aspect_ratio: 0.5625,
  file_path: '/image',
  height: 900,
  iso_639_1: null,
  vote_average: 123,
  vote_count: 123,
  width: 1600
}
