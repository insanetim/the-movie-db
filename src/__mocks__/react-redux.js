export const dispatch = jest.fn()

export const useDispatch = jest.fn(() => dispatch)
export const useSelector = jest.fn(fn => fn())
