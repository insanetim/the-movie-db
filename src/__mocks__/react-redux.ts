const dispatch = jest.fn()
const getState = jest.fn()

const useDispatch = jest.fn(() => dispatch)
const useSelector = jest.fn(fn => fn())

export { dispatch, getState, useDispatch, useSelector }
