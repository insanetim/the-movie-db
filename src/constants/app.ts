export const APP_NAME = 'THE MOVIE DB'
export const API_URL = 'https://api.themoviedb.org/3/'
export const ACCESS_TOKEN_AUTH =
  'eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiI1N2RiYjI4M2M4MDBjY2QwYmM5YTczNTJjN2Y1ZmEzYiIsInN1YiI6IjYzMmM4YThkZTNmYTJmMDA3ZmQ0ZjY5OSIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.nMVY53gL3_w8JAHI-N_uKQyXpvT9TsZ4_p8nhJSnAdQ'
export enum NOTIFICATION_TYPE {
  ERROR = 'error',
  SUCCESS = 'success'
}
export const NOTIFICATION_DURATION = 2.5
