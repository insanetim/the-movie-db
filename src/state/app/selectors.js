import { path } from 'ramda'

export const modalTypeSelector = path(['app', 'modal', 'modalType'])
export const modalPropsSelector = path(['app', 'modal', 'modalProps'])
export const notificationsSelector = path(['app', 'notifications'])
