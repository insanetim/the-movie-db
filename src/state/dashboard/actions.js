import * as types from './types'

export const fetchTrending = page => ({
  type: types.FETCH_TRENDING,
  payload: page
})

export const fetchTrendingRequest = () => ({
  type: types.FETCH_TRENDING_REQUEST
})

export const fetchTrendingSuccess = movies => ({
  type: types.FETCH_TRENDING_SUCCESS,
  payload: movies
})

export const fetchTrendingError = error => ({
  type: types.FETCH_TRENDING_ERROR,
  payload: error
})

export const setTrendingPage = page => ({
  type: types.SET_TRENDING_PAGE,
  payload: page
})

export const fetchSearch = ({ page, query }) => ({
  type: types.FETCH_SEARCH,
  payload: { page, query }
})

export const fetchSearchRequest = () => ({
  type: types.FETCH_SEARCH_REQUEST
})

export const fetchSearchSuccess = movies => ({
  type: types.FETCH_SEARCH_SUCCESS,
  payload: movies
})

export const fetchSearchError = error => ({
  type: types.FETCH_SEARCH_ERROR,
  payload: error
})

export const setSearchPage = page => ({
  type: types.SET_SEARCH_PAGE,
  payload: page
})
