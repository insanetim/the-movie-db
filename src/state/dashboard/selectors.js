import { path } from 'ramda'

export const trendingSelector = path(['dashboard', 'trending'])
export const searchSelector = path(['dashboard', 'search'])
