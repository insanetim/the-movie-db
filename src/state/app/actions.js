import { v4 as uuid } from 'uuid'

import { NOTIFICATION_DURATION, NOTIFICATION_TYPE } from 'src/constants'

import * as types from './types'

export const showModal = payload => ({
  type: types.SHOW_MODAL,
  payload
})

export const hideModal = () => ({
  type: types.HIDE_MODAL
})

export const showNotification = ({
  id = uuid(),
  messageType = NOTIFICATION_TYPE.SUCCESS,
  messageText,
  duration = NOTIFICATION_DURATION
}) => ({
  type: types.SHOW_NOTIFICATION,
  payload: {
    id,
    messageType,
    messageText,
    duration
  }
})

export const hideNotification = payload => ({
  type: types.HIDE_NOTIFICATION,
  payload
})
