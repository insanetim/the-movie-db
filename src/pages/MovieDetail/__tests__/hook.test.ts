import { act, renderHook } from '@testing-library/react'
import { useParams } from 'react-router-dom'
import { mockMovieDetail } from 'src/__mocks__/mockMovie'
import { dispatch } from 'src/__mocks__/react-redux'
import { showNotification } from 'src/store/app/actions'
import * as listsActions from 'src/store/lists/actions'
import * as movieActions from 'src/store/movie/actions'
import * as movieSelectors from 'src/store/movie/selectors'

import useContainer from '../hook'

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: jest.fn(() => ({ movieId: '123' }))
}))

jest.mock<typeof import('@reduxjs/toolkit')>('@reduxjs/toolkit', () => ({
  ...jest.requireActual('@reduxjs/toolkit'),
  nanoid: jest.fn(() => 'test/id')
}))

describe('MovieDetail useContainer hook', () => {
  const selectMovieById = jest
    .spyOn(movieSelectors, 'selectMovieById')
    .mockReturnValue(mockMovieDetail)

  it('should match snapshot', () => {
    const { result } = renderHook(useContainer)

    expect(result.current).toMatchSnapshot()
  })

  it('should match snapshot without params', () => {
    jest.mocked(useParams).mockReturnValueOnce({})
    const { result } = renderHook(useContainer)

    expect(result.current).toMatchSnapshot()
  })

  it('should check "handleFavoriteClick" method', () => {
    const changeMovieInFavorite = jest.spyOn(
      movieActions,
      'changeMovieInFavorite'
    )
    const notification = showNotification({
      messageText: 'test/title added to Favorite'
    })
    const { result } = renderHook(useContainer)

    act(() => {
      result.current.handleFavoriteClick()
    })

    expect(changeMovieInFavorite).toHaveBeenCalledWith({
      inFavorite: true,
      movieId: '123'
    })
    expect(dispatch).toHaveBeenCalledWith(notification)
  })

  it('should check "handleWatchlistClick" method', () => {
    const changeMovieInWatchlist = jest.spyOn(
      movieActions,
      'changeMovieInWatchlist'
    )
    const notification = showNotification({
      messageText: 'test/title added to Watchlist'
    })
    const { result } = renderHook(useContainer)

    act(() => {
      result.current.handleWatchlistClick()
    })

    expect(changeMovieInWatchlist).toHaveBeenCalledWith({
      inWatchlist: true,
      movieId: '123'
    })
    expect(dispatch).toHaveBeenCalledWith(notification)
  })

  it('should check "useEffect" method', () => {
    const fetchMovieDetail = jest.spyOn(movieActions, 'fetchMovieDetail')
    const fetchLists = jest.spyOn(listsActions, 'fetchLists')
    selectMovieById.mockReturnValueOnce(undefined)
    renderHook(useContainer)

    expect(dispatch).toHaveBeenCalled()
    expect(fetchMovieDetail).toHaveBeenCalledWith('123')
    expect(fetchLists).toHaveBeenCalledWith('1')
  })
})
