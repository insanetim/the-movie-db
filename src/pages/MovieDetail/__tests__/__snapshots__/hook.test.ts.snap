// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetail useContainer hook should match snapshot 1`] = `
{
  "error": null,
  "handleFavoriteClick": [Function],
  "handleWatchlistClick": [Function],
  "loading": false,
  "movie": {
    "accountStates": {
      "favorite": false,
      "id": 1234,
      "rated": false,
      "watchlist": false,
    },
    "adult": false,
    "backdrop_path": "/image",
    "belongs_to_collection": null,
    "budget": 1000000,
    "credits": {
      "cast": [
        {
          "adult": true,
          "cast_id": 1234,
          "character": "test/character",
          "credit_id": "123",
          "gender": null,
          "id": 1234,
          "known_for_department": "",
          "name": "test/name",
          "order": 1234,
          "original_name": "test/name",
          "popularity": 1234,
          "profile_path": "/image",
        },
      ],
      "crew": [
        {
          "adult": true,
          "credit_id": "123",
          "department": "test/department",
          "gender": null,
          "id": 1234,
          "job": "test/job",
          "known_for_department": "",
          "name": "test/name",
          "original_name": "test/name",
          "popularity": 1234,
          "profile_path": "/image",
        },
      ],
      "id": 1234,
    },
    "genre_ids": [
      1,
      2,
      3,
    ],
    "genres": [
      {
        "id": 1,
        "name": "test/genre",
      },
    ],
    "homepage": null,
    "id": 1234,
    "images": [
      {
        "aspect_ratio": 0.5625,
        "file_path": "/image",
        "height": 900,
        "iso_639_1": null,
        "vote_average": 1234,
        "vote_count": 1234,
        "width": 1600,
      },
    ],
    "imdb_id": null,
    "original_language": "en",
    "original_title": "test/title",
    "overview": "test/overview",
    "popularity": 1234,
    "poster_path": "/image",
    "production_companies": [
      {
        "id": 1234,
        "logo_path": null,
        "name": "test/company",
        "origin_country": "USA",
      },
    ],
    "production_countries": [
      {
        "iso_3166_1": "en",
        "name": "USA",
      },
    ],
    "release_date": "01-01-1999",
    "revenue": 3000000,
    "runtime": 150,
    "spoken_languages": [
      {
        "iso_639_1": "en",
        "name": "EN",
      },
    ],
    "status": "test/status",
    "tagline": null,
    "title": "test/title",
    "video": false,
    "vote_average": 1234,
    "vote_count": 1234,
  },
  "popoverOpen": false,
  "setPopoverOpen": [Function],
}
`;

exports[`MovieDetail useContainer hook should match snapshot without params 1`] = `
{
  "error": null,
  "handleFavoriteClick": [Function],
  "handleWatchlistClick": [Function],
  "loading": false,
  "movie": {
    "accountStates": {
      "favorite": false,
      "id": 1234,
      "rated": false,
      "watchlist": false,
    },
    "adult": false,
    "backdrop_path": "/image",
    "belongs_to_collection": null,
    "budget": 1000000,
    "credits": {
      "cast": [
        {
          "adult": true,
          "cast_id": 1234,
          "character": "test/character",
          "credit_id": "123",
          "gender": null,
          "id": 1234,
          "known_for_department": "",
          "name": "test/name",
          "order": 1234,
          "original_name": "test/name",
          "popularity": 1234,
          "profile_path": "/image",
        },
      ],
      "crew": [
        {
          "adult": true,
          "credit_id": "123",
          "department": "test/department",
          "gender": null,
          "id": 1234,
          "job": "test/job",
          "known_for_department": "",
          "name": "test/name",
          "original_name": "test/name",
          "popularity": 1234,
          "profile_path": "/image",
        },
      ],
      "id": 1234,
    },
    "genre_ids": [
      1,
      2,
      3,
    ],
    "genres": [
      {
        "id": 1,
        "name": "test/genre",
      },
    ],
    "homepage": null,
    "id": 1234,
    "images": [
      {
        "aspect_ratio": 0.5625,
        "file_path": "/image",
        "height": 900,
        "iso_639_1": null,
        "vote_average": 1234,
        "vote_count": 1234,
        "width": 1600,
      },
    ],
    "imdb_id": null,
    "original_language": "en",
    "original_title": "test/title",
    "overview": "test/overview",
    "popularity": 1234,
    "poster_path": "/image",
    "production_companies": [
      {
        "id": 1234,
        "logo_path": null,
        "name": "test/company",
        "origin_country": "USA",
      },
    ],
    "production_countries": [
      {
        "iso_3166_1": "en",
        "name": "USA",
      },
    ],
    "release_date": "01-01-1999",
    "revenue": 3000000,
    "runtime": 150,
    "spoken_languages": [
      {
        "iso_639_1": "en",
        "name": "EN",
      },
    ],
    "status": "test/status",
    "tagline": null,
    "title": "test/title",
    "video": false,
    "vote_average": 1234,
    "vote_count": 1234,
  },
  "popoverOpen": false,
  "setPopoverOpen": [Function],
}
`;
